keps/sig-storage/1710-selinux-relabeling/README.md

```diff
@@ -51,223 +48,183 @@
 
 See [SELinux documentation](https://selinuxproject.org/page/NB_MLS) for more details.
 
+
 ### SELinux context assignment
 In Kubernetes, the SELinux context of a pod is assigned in two ways:
 1. Either it is set by user in PodSpec or Container: https://kubernetes.io/docs/tasks/configure-pod-container/security-context/.
 1. If not set in Pod/Container, the container runtime will allocate a new unique SELinux context and assign it to a pod (container) by itself.
 
-### Volumes
-Currently, Kubernetes *knows*(*) which volume plugins supports SELinux (i.e. supports extended attributes on a filesystem the plugin provides).
-If SELinux is supported for a volume, it passes the volume to the container runtime with ":Z" option ("private unshared").
-The container runtime then **recursively relabels** all files on the volume to either the label set in PodSpec/Container or the random value allocated by the container runtime itself.
+### Volume mounting
+Linux kernel, with SELinux compiled in, allows `mount -o context=system_u:system_r:container_t:s0:c309,c383 <what> <where>` to mount a volume and pretend that all files on the volume have given SELinux label.
+
+Following conditions must be met:
+* It must be the first mount of the volume! It does not work for subsequent mounts of the volume, incl. bind-mounts.
+    * Second mount of the same volume fails with `mount point not mounted or bad option` when the `-o context=` does not match the first mount.
+    * A bind mount inherits `-o context=XYZ` option from the original mount (if it was set there). 
+* `/bin/mount` process must detect that the kernel supports SELinux. [It does so](https://github.com/util-linux/util-linux/blob/441f9b9303d015f1777aec7168807d58feacca31/libmount/src/context_mount.c#L291) by checking that both `/etc/selinux/config` file exists and `/sys/fs/selinux` is a mount point with "selinuxfs" type.
+    * `/bind/mount` throws away the SELinux mount options otherwise.
 
-**This relabeling needs to traverse through the whole volume, and it can be slow for volumes with a large amount of files.**
+Note that volumes mounted with `-o context` don't have `seclabel` in their mount options.
+In addition, calling `chcon` there will fail with `Operation not supported`.
+
+### SELinux support in volumes
+Currently, Kubernetes *knows*(*) which volume plugins support SELinux (i.e. supports extended attributes on a filesystem the plugin provides).
+If SELinux is supported for a volume, it passes the volume to the container runtime with ":Z" option ([`selinux_relabel`](https://github.com/kubernetes/cri-api/blob/648d24775c39780dfc367536f00197be64534684/pkg/apis/runtime/v1/api.proto#L205) in CRI).
+The container runtime then **recursively relabels** all files on the volume to either the label set in PodSpec/Container or the random value allocated by the container runtime itself.
 
 *) These in-tree volume plugins don't support SELinux: Azure File, CephFS, GlusterFS, HostPath, NFS, Portworx and Quobyte.
```

Comment:
Do we still need to mention the deprecated volume plugins, i.e., Quobyte?

Reply:
Quobyte removed.